core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -387,6 +399,208 @@ private Optional<Expression> unwrapTimestampToDateCast(Session session, Timestam
             };
         }
 
+        private Expression unwrapCast(BetweenPredicate expression)
+        {
+            // Canonicalization is handled by CanonicalizeExpressionRewriter
+            if (!(expression.getValue() instanceof Cast cast)) {
+                return expression;
+            }
+
+            Object rangeMin = new ExpressionInterpreter(expression.getMin(), plannerContext, session, typeAnalyzer.getTypes(session, types, expression.getMin()))
+                    .optimize(NoOpSymbolResolver.INSTANCE);
+            Object rangeMax = new ExpressionInterpreter(expression.getMax(), plannerContext, session, typeAnalyzer.getTypes(session, types, expression.getMax()))
+                    .optimize(NoOpSymbolResolver.INSTANCE);
+
+            if (rangeMin == null || rangeMin instanceof NullLiteral) {
+                return new Cast(new NullLiteral(), toSqlType(BOOLEAN));
+            }
+
+            if (rangeMin instanceof Expression || rangeMax instanceof Expression) {
+                return expression;
+            }
+
+            Type sourceType = typeAnalyzer.getType(session, types, cast.getExpression());
+            Type rangeMinType = typeAnalyzer.getType(session, types, expression.getMin());
+            Type rangeMaxType = typeAnalyzer.getType(session, types, expression.getMax());
```

Comment:
This looks like duplicating the logic quite much.

Did you try something like

```java
@Override
public Expression rewriteBetweenPredicate(BetweenPredicate node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
{
    BetweenPredicate expression = (BetweenPredicate) treeRewriter.defaultRewrite((Expression) node, null);

    ComparisonExpression lowBound = new ComparisonExpression(GREATER_THAN_OR_EQUAL, node.getValue(), node.getMin());
    ComparisonExpression highBound = new ComparisonExpression(LESS_THAN_OR_EQUAL, node.getValue(), node.getMax());

    Expression lowBoundUnwrapped = unwrapCast(lowBound);
    Expression highBoundUnwrapped = unwrapCast(highBound);
    if (lowBound.equals(lowBoundUnwrapped) && highBound.equals(highBoundUnwrapped)) {
        return expression;
    }
    ...
}
```

?
Why it cannot be made to work?

Reply:
Created https://github.com/trinodb/trino/pull/14648 with inspiration from https://github.com/trinodb/trino/pull/12797

I stumbled while trying to add the methods:

- `getPreviousValue(Object)`
- `getNextValue(Object)`

to `DateType`.
I am assuming that checking the validity of a date needs joda-time expertise which is not available in `trino-spi` module.

As an alternative we could use custom operators included in `DateOperators` to cover this need.

Reply:
> I stumbled while trying to add the methods:
> 
> * `getPreviousValue(Object)`
> * `getNextValue(Object)`
> 
> to `DateType`.

This is now covered by https://github.com/trinodb/trino/pull/12797.